text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+        //  -   Access to the next `mem::size_of::<T>() - old_layout.size()` bytes is exclusive.
+        unsafe { ptr::write_bytes(pointer, 0, mem::size_of::<T>() - old_layout.size()) };
+
+        Ok(((), mem::size_of::<T>()))
+    }
+}
+
+//  Safety:
+//  -   `self.resolve(handle)` always returns the same address, as long as `self` doesn't move.
+unsafe impl<T> StableStore for InlineSingleStore<T> {}
+
+impl<T> fmt::Debug for InlineSingleStore<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        let layout = Layout::new::<T>();
+
+        f.debug_struct("InlineSingleStore")
+            .field("size", &layout.size())
+            .field("align", &layout.align())
+            .finish()
+    }
+}
+
+impl<T> InlineSingleStore<T> {
+    fn validate_layout(layout: Layout) -> Result<(), AllocError> {
+        let own = Layout::new::<T>();
+
+        if layout.align() <= own.align() && layout.size() <= own.size() {
+            Ok(())
+        } else {
+            Err(AllocError)
+        }
+    }
+}
+```
+
+And that's it!
+
+I need not implement `MultipleStore`, and thus do not have to track allocations and deallocations. And I need not
+implement `PinningStore`, and thus do not have to ensure memory address stability across moves.
+
+More examples of `Store` can be found at https://github.com/matthieu-m/storage/tree/main/src/store, including an inline
+bump store.
+
+
+#   Reference-level explanation
+
+##  Overview
+
+This RFC introduces 4 new traits.
+
+The core of this RFC is the `Store` trait:
+
+```rust
+/// Allocates and deallocates handles to blocks of memory, which can be temporarily resolved to pointers to actually
+/// access said memory.
+pub unsafe trait Store {
+    /// Handle to a block of memory.
+    type Handle: Copy;
+
+    /// Returns a dangling handle, always invalid.
+    fn dangling() -> Self::Handle;
+
+    /// Return a pointer to the block of memory associated to `handle`.
+    unsafe fn resolve(&self, handle: Self::Handle) -> NonNull<u8>;
+
+    //  The following methods are similar to `Allocator`, reformulated in terms of `Self::Handle`.
+
+    /// Allocates a new handle to a block of memory. On success, invalidates any existing handle.
+    fn allocate(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Deallocates a handle.
+    unsafe fn deallocate(&self, handle: Self::Handle, layout: Layout);
+
+    /// Grows the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn grow(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Shrinks the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn shrink(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
```

Comment:
Not sure, to be honest. I just decided to fully mirror the `Allocator` API as a starting point.

In the companion repository, the `InlineBumpStore` takes advantage of the split by having `shrink` be a no-op, but even that is questionable -- it could in theory shrink the last allocated memory block if needed.